pulsar-client-go/pulsar/schema.go

```diff
@@ -0,0 +1,494 @@
+	si := &SchemaInfo{
+		Name:   name,
+		Schema: schema,
+		Type:   schemaType,
+	}
+	return si
+}
+
+type Schema interface {
+	Serialize(v interface{}) ([]byte, error)
+	UnSerialize(data []byte, v interface{}) error
+	Validate(message []byte) error
+	GetSchemaInfo() *SchemaInfo
+}
+
+type SchemaDefinition struct {
+	SchemaDef *goavro.Codec
+}
+
+func NewSchemaDefinition(schema *goavro.Codec) *SchemaDefinition {
+	schemaDef := &SchemaDefinition{
+		SchemaDef: schema,
+	}
+	return schemaDef
+}
+
+// initCodec returns a Codec used to translate between a byte slice of either
+// binary or textual Avro data and native Go data.
+func initCodec(codec string) (*goavro.Codec, error) {
+	return goavro.NewCodec(codec)
+}
+
+type JsonSchema struct {
+	SchemaDefinition
+	SchemaInfo
+}
+
+func NewJsonSchema(codec string) *JsonSchema {
+	js := new(JsonSchema)
+	schema, err := initCodec(codec)
+	if err != nil {
+		log.Fatalf("init codec error:%v", err)
+	}
+	schemaDef := NewSchemaDefinition(schema)
+	js.SchemaInfo.Schema = schemaDef.SchemaDef.Schema()
+	js.SchemaInfo.Type = JSON
+	return js
+}
+
+func (js *JsonSchema) Serialize(data interface{}) ([]byte, error) {
+	return json.Marshal(data)
+}
+
+func (js *JsonSchema) UnSerialize(data []byte, v interface{}) error {
+	return json.Unmarshal(data, v)
+}
+
+func (js *JsonSchema) Validate(message []byte) error {
+	return js.UnSerialize(message, nil)
+}
+
+func (js *JsonSchema) GetSchemaInfo() *SchemaInfo {
+	jsonSchema := NewSchemaInfo("Json", js.SchemaInfo.Schema, JSON)
+	return jsonSchema
+}
+
+type ProtoSchema struct {
+	SchemaDefinition
+	SchemaInfo
+}
+
+func NewProtoSchema(codec string) *ProtoSchema {
+	ps := new(ProtoSchema)
+	schema, err := initCodec(codec)
+	if err != nil {
+		log.Fatalf("init codec error:%v", err)
+	}
+	schemaDef := NewSchemaDefinition(schema)
+	ps.SchemaDefinition.SchemaDef = schemaDef.SchemaDef
+	ps.SchemaInfo.Schema = schemaDef.SchemaDef.Schema()
+	ps.SchemaInfo.Type = PROTOBUF
+	return ps
+}
+
+func (ps *ProtoSchema) Serialize(data interface{}) ([]byte, error) {
+	return proto.Marshal(data.(proto.Message))
+}
+
+func (ps *ProtoSchema) UnSerialize(data []byte, v interface{}) error {
+	return proto.Unmarshal(data, v.(proto.Message))
+}
+
+func (ps *ProtoSchema) Validate(message []byte) error {
+	return ps.UnSerialize(message, nil)
+}
+
+func (ps *ProtoSchema) GetSchemaInfo() *SchemaInfo {
+	jsonSchema := NewSchemaInfo("Proto", ps.SchemaDef.Schema(), PROTOBUF)
+	return jsonSchema
+}
+
+type AvroSchema struct {
+	SchemaDefinition
+	SchemaInfo
+}
+
+func NewAvroSchema(codec string) *AvroSchema {
+	as := new(AvroSchema)
+	schema, err := initCodec(codec)
+	if err != nil {
+		log.Fatalf("init codec error:%v", err)
+	}
+	schemaDef := NewSchemaDefinition(schema)
+	as.SchemaDefinition.SchemaDef = schemaDef.SchemaDef
+	as.SchemaInfo.Schema = schemaDef.SchemaDef.Schema()
+	as.SchemaInfo.Type = AVRO
+	return as
+}
+
+func (as *AvroSchema) Serialize(data interface{}) ([]byte, error) {
```

Comment:
this serialization here doesn't seem to be correct to me. you don't need line 188-197. you can just simply call `BinaryFromNative`.